### PR TITLE
feat: improve StdChains

### LINKED ISSUE #232: `[rpc_endpoints]` vars should be lazily loaded
By default in forge, `[rpc_endpoints]` entries are loaded lazily. So if it contains e.g. `goerli = ${GOERLI_URL}` but the `GOERLI_URL` is not defined, that's fine as long as goerli isn't used by a test/script.

Whatever the wisdom of doing that, forge-std breaks this since #195. Maybe moving `stdChains` from a mapping to a function that does the custom-or-default logic on the fly would be as good?

Some context to the use case: a foundry repo used a dependency by a non-foundry repo. Some scripts of the foundry repo are used, but the networks never are.

### PR DESCRIPTION
**Pull request by adhusson**
* Lazy-load `[rpc_endpoints]` entries as needed [1] (closes #232 for real)
* `rpcUrl` can be set as 1) custom, 2) config entry, 3) default value
* No empty chainAlias, no 0 chainId, clearer error messages
* Shorter via_ir compilation time (on my machine, 68s -> 42s with default runs)
* Remove extraneous keccak256 calls

[1] forge lazy-loads `[rpc_endpoints]`. This is useful when e.g. CI is only running tests that use the `TESTNET_URL` env var, but don't know about e.g. `MAINNET_URL` and `FLASHBOTS_URL`. forge-std iterating through all rpc urls breaks this.

### CODE REVIEW

**Comment by ZeroEkkusu** [comment]
Haven't reviewed;

> don't know about e.g. MAINNET_URL and FLASHBOTS_URL

Sorry, could you explain to me or give me an example of the problem? I see the code is more complicated.

> Shorter via_ir compilation time (on my machine, 68s -> 42s with default runs)

I saw that it uses the `_constructor` again, so I tested it on Uniswap/permit2. It takes 200s vs the current 140s. Out checks also show longer `via-ir` compilation times (9s vs 3s).

**Comment by adhusson** [comment]
Thanks for taking a look!

re: 1, suppose you have
```toml
[rpc_endpoints]
testnet="${TESTNET_URL}"
mainnet="${MAINNET_URL}"
```

With foundry only, if a consumer of the project only knows about `TESTNET_URL` or `MAINNET_URL` (a CI, or a client library calling scripts), that's fine. The entry with an undefined var will never be interpreted and never cause an error. With forge-std, `rpcUrls` is called whenever some chain info is accessed, which throws if at least one `[rpc_endpoints]` entry cannot be read. Basically the gains of lazy-loading are gone.

re: 2, I confirm that compiling permit2 with `_constructor` is slower. I've just brought back the `initialize()` function.

File: src/StdChains.sol
Comment (adhusson) [reply src/StdChains.sol:83]: `chain` is a memory reference. All modifications to it will be visible throughout the current EVM call frame. Not restoring the `rpcUrl` means that `setChainWithDefaultRpcUrl` sets its argument's `rpcUrl` property to the empty.

File: src/StdChains.sol
Comment (adhusson) [reply src/StdChains.sol:78]: Same, I see myself using that in toolkit contracts that inherit from `StdChains`.

File: src/StdChains.sol
Comment (adhusson) [reply src/StdChains.sol:27]: People who extend `StdChains`/forge-std with their own tooling may want to customize how chains are handled.

File: src/StdChains.sol
Comment (adhusson) [reply src/StdChains.sol:58]: `chain` is a memory reference, so the argument passed is just a memory offset. No copying occurs.

**Review by ZeroEkkusu — CHANGES_REQUESTED** [review]
I need to test this locally, but here's what I've noticed:

- I don't think the `memory chain` changes persist
- Reverts should says `uint256` instead of `uint`
- `setChain` doesn't call `initialize` so it wouldn't revert if you tried setting a Chain with chain ID 1 immediately
- Same as previous, you're able to set a Chain with chain ID 1, and if you then called `getChain("mainnet")` it would revert because a chain with chain ID 1 already exisits, which is confusing

File: src/StdChains.sol
Comment (ZeroEkkusu) [reply src/StdChains.sol:27]: Let's do `_*` then - e.g. `_chains`.

File: src/StdChains.sol
Comment (ZeroEkkusu) [reply src/StdChains.sol:81]: Same as above.

**Comment by adhusson** [comment]
Thanks for the comments, I've applied the changes. For the `memory chain`, you can try by adding this test

```solidity
function testPersist() public {
  Chain memory chain = Chain("a",123,"URL");
  console.log("before",chain.rpcUrl);
  setChainWithDefaultRpcUrl("alias",chain);
  console.log("after",chain.rpcUrl);
  }
```
and then toggle that line and see the difference:
https://github.com/foundry-rs/forge-std/blob/d65123c8354ca60019442d9374a9b0acd8cfad43/src/StdChains.sol#L86

File: src/StdChains.sol
Comment (ZeroEkkusu) [reply src/StdChains.sol:81]: @adhusson - you didn't add an underscore before the name.

**Comment by ZeroEkkusu** [comment]
Good to merge after the discussion is resolved.

File: src/Script.sol
Comment (mds1) [inline src/Script.sol:3]: Is the thinking here that we add this to `Script.sol` and `Test.sol` so users on <0.8 don't need to add it themselves? If that seems fine, and my only question is: ABI encoder v2 is the default in 0.8, so does explicitly specifying it here as experimental change anything? (i.e. like using an older/buggier version of the v2 encoder or something)

File: src/StdChains.sol
Comment (mds1) [inline src/StdChains.sol:28]: Let's add the comments back to keep these a bit more self-documenting for future contributors
```suggestion
    // Maps from a chain ID to that chain's alias
    mapping(uint256 => string) internal _idToAlias;
```

File: src/StdChains.sol
Comment (mds1) [inline src/StdChains.sol:27]: Similarly
(matching the alias in the `foundry.toml` file) to chain data.
```suggestion
    // Maps from the chain's alias (matching the alias in the `foundry.toml` file) to chain data.
    mapping(string => Chain) internal _chains;
    // Maps from the chain's alias (matching the alias in the `foundry.toml` file) to it's default RPC URL
    mapping(string => string) internal _defaultRpcUrls;
```

File: test/StdChains.t.sol
Comment (mds1) [inline test/StdChains.t.sol:20]: We now only check an arbitrary subset of default URLs, but we should still test all our default RPCs. Can conditionally skip any aliases that we know have invalid ones and will revert

File: src/StdChains.sol
Comment (mds1) [inline src/StdChains.sol:122]: @adhusson @ZeroEkkusu can we get some code comments in this PR, and maybe after some book updates, for StdChains / this PR? There's now a lot going on in `StdChains` and it's fairly complex, so the UX isn't the most intuitive by skimming the code (and there's a lot to skim). To start, a natspec block right above the `abstract contract StdChains {` line would be helpful. Some questions I'd like answered by those comments:
1. What's the default, out of the box behavior when I:
    1. Access a chain with a pre-existing alias, like "mainnet"
    2. Access a chain without a pre-existing alias, like "myNewChain"
    3.  For (1) and (2) how does it vary if nothing is in my foundry.toml vs. if I have a hardcoded RPC URL vs. if I have an env var, and that env var does or does not exist
2. How do I add a new chain?
3. What's the difference between `setChain` and `setChainWithDefaultRpcUrl` and when do I use each?
4. How do I rename an existing chain, like changing `anvil` to `hardhat`?

File: src/StdChains.sol
Comment (mds1) [inline src/StdChains.sol:56]: Why do we call a view method at the end of this function? If that's intentional/required, def not intuitive and we should document what's going on here with some comments

Comment (mds1) [reply src/StdChains.sol:56]: Seems like you're modifying the `chain` memory var in `withRpcUrl` and then it bubbles up to here so that new RP URL is returned. But that's a bit strange, why not just return the new RPC URL and set it explicitly?

File: src/StdChains.sol
Comment (mds1) [inline src/StdChains.sol:61]: `withRpcUrl` feels like a bad name. maybe `findRpcUrl` or `getRpcUrl`, then return the RPC URL and set it in the parent method

File: src/Script.sol
Comment (ZeroEkkusu) [reply src/Script.sol:3]: No, I planned to remove this along with `Components`.

File: src/StdChains.sol
Comment (ZeroEkkusu) [reply src/StdChains.sol:28]: Do we need a more discouraging naming, so that users don’t shout themselves in the foot? E.g. `_id_to_alias_`.

This is only for devs.

File: src/StdChains.sol
Comment (ZeroEkkusu) [reply src/StdChains.sol:122]: Btw this function should also be only for devs and have the same discouraging naming style.

File: src/StdChains.sol
Comment (mds1) [reply src/StdChains.sol:28]: What about just making it `private` instead of a weird naming style?

Comment (ZeroEkkusu) [reply src/StdChains.sol:28]: https://github.com/foundry-rs/forge-std/pull/247#discussion_r1043412377 - def not the best solution - if @adhusson doesn’t intend to extend the features immediately, a better solution is to keep everything private, as was, and if someone requests it in the future, make changes then. What do you think?

Comment (mds1) [reply src/StdChains.sol:28]: Agreed, let's keep as much of the internal implementation `private` as possible to keep the UX/interface as simple/safe as possible, and we can iterate in the future when the need arises

**Comment by ZeroEkkusu** [comment]
Pushed changes for `1.2.0` release.

After [implemet](https://github.com/foundry-rs/foundry/issues/3871) is resolved, we should be able to revert the changes without changing the API.

File: src/StdChains.sol
Comment (mds1) [reply src/StdChains.sol]: Done in 906e150

**Comment by mds1** [comment]
Still a few open comments here to address before merging

File: src/StdChains.sol
Comment (ZeroEkkusu) [reply src/StdChains.sol:118]: There are 3 functions: 2 `getChain` and 1 `setChain`. The first time any one of these is called, chains are initialized with the default set of RPC URLs. This is done by `initialize`, which uses `setChainWithDefaultRpcUrl`. Defaults are recorded in `defaultRpcUrls`.

`setChain` is straightforward. It just sets a chain (in `chains`).

`getChain` uses `withRpcUrl` to return a chain. For example, let's say we want to retrieve `mainnet`. If you haven't set it with `setChain` and you haven't specified it in `foundry.toml`, the default one will be returned. If you set it in `foundry.toml` it will return that, if valid (e.g. the env exists), otherwise, the default one. If you specified it with `setChain` it will return that.

Comment (mds1) [reply src/StdChains.sol:147]: Thanks for the docs, I modified them slightly and ported them as comments in 4b9b197. We can pull that and reuse it for the book PR too

**Review by ZeroEkkusu — APPROVED** [review]

**Pull request merged**
